Sources/Machine/rbm_spin_v2.cc

```diff
@@ -0,0 +1,325 @@
+#include "Utils/pybind_helpers.hpp"
+
+namespace netket {
+
+RbmSpinV2::RbmSpinV2(std::shared_ptr<const AbstractHilbert> hilbert,
+                     Index nhidden, Index alpha, bool usea, bool useb,
+                     Index const batch_size)
+    : AbstractMachine{std::move(hilbert)},
+      W_{},
+      a_{nonstd::nullopt},
+      b_{nonstd::nullopt},
+      theta_{} {
+  const auto nvisible = GetHilbert().Size();
+  assert(nvisible >= 0 && "AbstractHilbert::Size is broken");
+  if (nhidden < 0) {
+    std::ostringstream msg;
+    msg << "invalid number of hidden units: " << nhidden
+        << "; expected a non-negative number";
+    throw InvalidInputError{msg.str()};
+  }
+  if (alpha < 0) {
+    std::ostringstream msg;
+    msg << "invalid density of hidden units: " << alpha
+        << "; expected a non-negative number";
+    throw InvalidInputError{msg.str()};
+  }
+  if (nhidden > 0 && alpha > 0 && nhidden != alpha * nvisible) {
+    std::ostringstream msg;
+    msg << "number and density of hidden units are incompatible: " << nhidden
+        << " != " << alpha << " * " << nvisible;
+    throw InvalidInputError{msg.str()};
+  }
+  nhidden = std::max(nhidden, alpha * nvisible);
+
+  W_.resize(nvisible, nhidden);
+  if (usea) {
+    a_.emplace(nvisible);
+  }
+  if (useb) {
+    b_.emplace(nhidden);
+  }
+
+  theta_.resize(batch_size, nhidden);
+}
+
+Index RbmSpinV2::BatchSize() const noexcept { return theta_.rows(); }
+
+void RbmSpinV2::BatchSize(Index batch_size) {
+  if (batch_size <= 0) {
+    std::ostringstream msg;
+    msg << "invalid batch size: " << batch_size
+        << "; expected a positive number";
+    throw InvalidInputError{msg.str()};
+  }
+  if (batch_size != BatchSize()) {
+    theta_.resize(batch_size, theta_.cols());
+  }
+}
+
+Eigen::VectorXcd RbmSpinV2::GetParameters() {
+  Eigen::VectorXcd parameters(Npar());
+  Index i = 0;
+  if (a_.has_value()) {
+    parameters.segment(i, a_->size()) = *a_;
+    i += a_->size();
+  }
+  if (b_.has_value()) {
+    parameters.segment(i, b_->size()) = *b_;
+    i += b_->size();
+  }
+  parameters.segment(i, W_.size()) =
+      Eigen::Map<Eigen::VectorXcd>(W_.data(), W_.size());
+  return parameters;
+}
+
+void RbmSpinV2::SetParameters(Eigen::Ref<const Eigen::VectorXcd> parameters) {
+  if (parameters.size() != Npar()) {
+    std::ostringstream msg;
+    msg << "wrong shape: [" << parameters.size() << "]; expected [" << Npar()
+        << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  Index i = 0;
+  if (a_.has_value()) {
+    *a_ = parameters.segment(i, a_->size());
+    i += a_->size();
+  }
+  if (b_.has_value()) {
+    *b_ = parameters.segment(i, b_->size());
+    i += b_->size();
+  }
+  Eigen::Map<Eigen::VectorXcd>(W_.data(), W_.size()) =
+      parameters.segment(i, W_.size());
+}
+
+void RbmSpinV2::LogVal(Eigen::Ref<const RowMatrix<double>> x,
+                       Eigen::Ref<Eigen::VectorXcd> out, const any &) {
+  if (x.cols() != Nvisible()) {
```

Comment:
Worth moving these checks in a recyclable function to be called from derivatives too 
[I see you have done this for other checks, I guess you will do it here too]